dapps-certification-persistence/src/IOHK/Certification/Persistence/API.hs

```diff
@@ -1,16 +1,89 @@
+{-# LANGUAGE OverloadedLabels    #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
 {-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE OverloadedStrings #-}
-{-# LANGUAGE OverloadedLabels #-}
-{-# LANGUAGE RecordWildCards #-}
-{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeOperators       #-}
 
 module IOHK.Certification.Persistence.API where
 
-import Database.Selda
-import Database.Selda.SQLite
-import IOHK.Certification.Persistence.Structure
-import Data.Maybe
-import Control.Monad
+import           Control.Monad
+import           Data.Maybe
+import           Database.Selda
+import           Database.Selda.SQLite
+import           IOHK.Certification.Persistence.Structure
+
+getTransactionIdQ:: Text -> Query t (Col t (ID Transaction))
+getTransactionIdQ  externalAddress = do
+  p <- select transactions
+  restrict (p ! #wtxExternalId .== text externalAddress)
+  pure (p ! #wtxId)
+
+getTransactionId :: MonadSelda m => Text -> m (Maybe (ID Transaction))
+getTransactionId = fmap listToMaybe . query . getTransactionIdQ
+
+-- | inserts or updates a transaction
+-- NOTE: if the transaction exists Nothing will be returned and entries
+-- will not be updated
```

Comment:
Then this should be call `insert` not `upsert`, right? `upsert` implies I can update something that already exists.

Reply:
Oh I see we can update the transaction but not its entries